streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -388,76 +392,107 @@ public void process(final ProcessorSupplier<K, V> processorSupplier, String... s
             Serde<V1> otherValueSerde,
             boolean outer) {
 
-        Set<String> allSourceNodes = ensureJoinableWith((AbstractStream<K>) other);
-
-        StateStoreSupplier thisWindow = Stores.create(windows.name() + "-this")
-                .withKeys(keySerde)
-                .withValues(thisValueSerde)
-                .persistent()
-                .windowed(windows.maintainMs(), windows.segments, true)
-                .build();
-
-        StateStoreSupplier otherWindow = Stores.create(windows.name() + "-other")
-                .withKeys(keySerde)
-                .withValues(otherValueSerde)
-                .persistent()
-                .windowed(windows.maintainMs(), windows.segments, true)
-                .build();
+        return doJoin(other,
+                      joiner,
+                      windows,
+                      keySerde,
+                      thisValueSerde,
+                      otherValueSerde,
+                      new DefaultJoin(outer));
+    }
+
+    private <V1, R> KStream<K, R> doJoin(KStream<K, V1> other,
+                                         ValueJoiner<V, V1, R> joiner,
+                                         JoinWindows windows,
+                                         Serde<K> keySerde,
+                                         Serde<V> thisValueSerde,
+                                         Serde<V1> otherValueSerde,
+                                         KStreamImplJoin join) {
+        KStreamImpl<K, V> joinThis = this;
+        KStreamImpl<K, V1> joinOther = (KStreamImpl) other;
+
+        if (joinThis.repartitionRequired) {
+            joinThis = joinThis.repartitionForJoin(keySerde, thisValueSerde, joinOther
+                .repartitionRequired);
+        }
 
-        KStreamJoinWindow<K, V> thisWindowedStream = new KStreamJoinWindow<>(thisWindow.name(), windows.before + windows.after + 1, windows.maintainMs());
-        KStreamJoinWindow<K, V1> otherWindowedStream = new KStreamJoinWindow<>(otherWindow.name(), windows.before + windows.after + 1, windows.maintainMs());
+        if (joinOther.repartitionRequired) {
+            joinOther = joinOther.repartitionForJoin(keySerde, otherValueSerde, this
+                .repartitionRequired);
+        }
 
-        KStreamKStreamJoin<K, R, V, V1> joinThis = new KStreamKStreamJoin<>(otherWindow.name(), windows.before, windows.after, joiner, outer);
-        KStreamKStreamJoin<K, R, V1, V> joinOther = new KStreamKStreamJoin<>(thisWindow.name(), windows.before, windows.after, reverseJoiner(joiner), outer);
+        // if both input streams need repartitioning we need to make sure
+        // that they are co-partitioned as this means the repartitioned streams
+        // will also be co-partitioned
```

Comment:
I'm still not totally clear about this motivation, so let me just elaborate my thoughts so that we can discuss further. This is related to my other comments regarding co-partitioning as well. I'll use the following running example:

```
stream1 = builder.stream("a").selectKey("C");   // topic "a" has 4 partitions, keyed on "A";
stream2 = builder.stream("b").selectKey("C");   // topic "b" has 5 partitions, keyed on "B";

stream3 = stream1.join(stream2);
```

1. Today when the library auto creates a re-partition topic, it effectively is splitting a user-defined topology into two sub-topologies, separated by this topic. So in the above case, two repartition topics (let's call them "a-repartition" and "b-repartition") are created which separate the topology into three sub-topologies:

```
stream1 = builder.stream("a").selectKey("C");   // topic "a" has 4 partitions, keyed on "A";
stream1.to("a-repartition");         // topic "a-repartition" is keyed on "C" 
```

```
stream2 = builder.stream("b").selectKey("C");   // topic "b" has 5 partitions, keyed on "B";
stream2.to("b-repartition");         // topic "b-repartition" is keyed on "C" 
```

```
stream1-rep = builder.stream("a-repartition");   // topic "a-repartition" has 5 partitions, keyed on "C";
stream2-rep = builder.stream("b-repartition");   // topic "b-repartition" has 5 partitions, keyed on "C";
stream1-rep.join(stream2-rep);
```

2. With the default `PartitionGrouper`, the first sub-topology will have 4 tasks, and the second sub-topology will have 5 tasks. And hence WITHOUT considering the third sub-topology, the auto created topic should have number partitions equal to 4 and 5 respectively even without calling `copartition`. This logic is written in `StreamsPartitionAssignor.assign`, looking for comments

```
        // for those internal source topics that do not have co-partition enforcement,
        // set the number of partitions to the maximum of the depending sub-topologies source topics
```

3. Now with the third sub-topology, we need to make sure the re-partition topics `a-repartition` and `b-repartition` needs to be co-partitioned, which can be achieved by calling `joinThis.ensureJoinableWith(joinOther);`, that as a result they should all have 5 partitions. But the source topics `a` and `b` do NOT need to be co-partitioned.

4. Now assuming we change the above example with 

```
stream0 = builder.stream("z");   // keyed on "A", with 3 partitions;
stream1 = builder.stream("a").merge(stream0);    // keyed on "A", with 4 partitions;

// same as before, only difference now is that first sub-topology now has two source topics with different number of partitions
```

Auto creating the repartition topic `a-repartition` should NOT require `a` and `z` to be co-partitioned.